#!/usr/bin/env python3
"""
test_server.py â€” Skynet æ¸¸æˆæœåŠ¡å™¨ WebSocket æµ‹è¯•è„šæœ¬

ä½¿ç”¨æ–¹æ³•:
    python test_server.py                       # è¿è¡Œå…¨éƒ¨æµ‹è¯• (é»˜è®¤ ws://127.0.0.1:9948)
    python test_server.py --host 192.168.1.10   # æŒ‡å®šä¸»æœº
    python test_server.py --port 9949           # æŒ‡å®šç«¯å£
    python test_server.py --test register       # åªè·‘æ³¨å†Œæµ‹è¯•
    python test_server.py --test login           # åªè·‘ç™»å½•æµ‹è¯•
    python test_server.py --test heartbeat       # åªè·‘å¿ƒè·³æµ‹è¯•
    python test_server.py --test reconnect       # åªè·‘é¡¶å·/é‡è¿æµ‹è¯•
    python test_server.py --test stress          # å‹åŠ›æµ‹è¯•(å¤šè¿æ¥)
    python test_server.py --test all             # å…¨éƒ¨æµ‹è¯•(é»˜è®¤)

ä¾èµ–:
    pip install websockets protobuf grpcio-tools

Proto ç¼–è¯‘(é¦–æ¬¡):
    python -m grpc_tools.protoc -I. --python_out=. Game.proto
"""

import asyncio
import argparse
import struct
import time
import sys
import os
import traceback

import websockets

# â”€â”€ Proto ç¼–è¯‘ & å¯¼å…¥ â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
# å°è¯•å¯¼å…¥ç¼–è¯‘åçš„ pb æ–‡ä»¶ï¼Œè‹¥ä¸å­˜åœ¨åˆ™å½“åœºç¼–è¯‘
try:
    import Game_pb2 as pb
except ImportError:
    # è·å–å½“å‰è„šæœ¬æ‰€åœ¨ç›®å½• (Test ç›®å½•)
    current_dir = os.path.dirname(os.path.abspath(__file__))
    # è·å–é¡¹ç›®æ ¹ç›®å½• (server ç›®å½•)
    root_dir = os.path.dirname(current_dir)
    # å®šä¹‰ Proto æ–‡ä»¶å¤¹è·¯å¾„
    proto_dir = os.path.join(root_dir, "Proto")
    proto_file = os.path.join(proto_dir, "Game.proto")

    if os.path.exists(proto_file):
        print(f"[*] Compiling {proto_file} ...")
        from grpc_tools import protoc
        # æ ¸å¿ƒä¿®å¤ï¼š-I å¿…é¡»æŒ‡å‘åŒ…å« proto æ–‡ä»¶çš„ç›®å½•
        protoc.main([
            "grpc_tools.protoc",
            f"-I{proto_dir}", 
            f"--python_out={current_dir}", 
            proto_file
        ])
        
        # ç¡®ä¿ç¼–è¯‘åçš„æ–‡ä»¶èƒ½è¢« import
        sys.path.append(current_dir)
        import Game_pb2 as pb
        print("[*] Compiled successfully.")
    else:
        print(f"[!] Game.proto not found at {proto_file}")
        sys.exit(1)


# â”€â”€ åè®®å·(ä¸ Proto/MsgId.lua ä¸€è‡´) â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
class MsgId:
    C2S_Login          = 1001
    S2C_LoginResult    = 1002
    C2S_Register       = 1003
    S2C_RegisterResult = 1004
    C2S_Logout         = 1101
    S2C_Kick           = 1102
    C2S_JoinRoom       = 2001
    S2C_JoinResult     = 2002
    C2S_RoomAction     = 2003
    S2C_RoomSync       = 2004
    C2S_Ping           = 9001
    S2C_Pong           = 9002


# â”€â”€ msgId -> protobuf æ¶ˆæ¯ç±» æ˜ å°„ â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
ENCODE_MAP = {
    MsgId.C2S_Login:      pb.C2S_Login,
    MsgId.C2S_Register:   pb.C2S_Register,
    MsgId.C2S_Logout:     pb.C2S_Logout,
    MsgId.C2S_JoinRoom:   pb.C2S_JoinRoom,
    MsgId.C2S_RoomAction: pb.C2S_RoomAction,
    MsgId.C2S_Ping:       pb.C2S_Ping,
}

DECODE_MAP = {
    MsgId.S2C_LoginResult:    pb.S2C_LoginResult,
    MsgId.S2C_RegisterResult: pb.S2C_RegisterResult,
    MsgId.S2C_Kick:           pb.S2C_Kick,
    MsgId.S2C_JoinResult:     pb.S2C_JoinResult,
    MsgId.S2C_RoomSync:       pb.S2C_RoomSync,
    MsgId.S2C_Pong:           pb.S2C_Pong,
}

MSGID_NAME = {v: k for k, v in vars(MsgId).items() if isinstance(v, int)}


# â”€â”€ ç¼–è§£ç  â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
def encode(msg_id: int, **fields) -> bytes:
    """ç¼–ç : [2B msgId big-endian][protobuf payload]"""
    cls = ENCODE_MAP.get(msg_id)
    if cls is None:
        raise ValueError(f"Unknown encode msgId: {msg_id}")
    msg = cls(**fields)
    payload = msg.SerializeToString()
    header = struct.pack("!H", msg_id)
    return header + payload


def decode(data: bytes):
    """è§£ç : è¿”å› (msgId, protobuf message object)"""
    if len(data) < 2:
        return None, None
    msg_id = struct.unpack("!H", data[:2])[0]
    cls = DECODE_MAP.get(msg_id)
    if cls is None:
        return msg_id, None
    msg = cls()
    msg.ParseFromString(data[2:])
    return msg_id, msg


# â”€â”€ å®¢æˆ·ç«¯å°è£… â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
class GameClient:
    """è½»é‡ WebSocket å®¢æˆ·ç«¯ï¼Œå°è£…æ”¶å‘é€»è¾‘"""

    def __init__(self, uri: str, name: str = "Client"):
        self.uri = uri
        self.name = name
        self.ws = None
        self.uid = None

    async def connect(self):
        self.ws = await websockets.connect(self.uri, max_size=1 << 20)
        log(self.name, "connected")

    async def close(self):
        if self.ws:
            await self.ws.close()
            log(self.name, "disconnected")

    async def send(self, msg_id: int, **fields):
        data = encode(msg_id, **fields)
        await self.ws.send(data)
        log(self.name, f">>> {MSGID_NAME.get(msg_id, msg_id)}  {fields}")

    async def recv(self, timeout: float = 5.0):
        """æ¥æ”¶ä¸€æ¡æ¶ˆæ¯ï¼Œè¶…æ—¶æŠ›å¼‚å¸¸"""
        raw = await asyncio.wait_for(self.ws.recv(), timeout=timeout)
        msg_id, msg = decode(raw)
        name = MSGID_NAME.get(msg_id, str(msg_id))
        fields = {f.name: getattr(msg, f.name) for f in msg.DESCRIPTOR.fields} if msg else {}
        log(self.name, f"<<< {name}  {fields}")
        return msg_id, msg

    async def recv_optional(self, timeout: float = 1.0):
        """å°è¯•æ¥æ”¶ï¼Œè¶…æ—¶è¿”å› (None, None)"""
        try:
            return await self.recv(timeout=timeout)
        except (asyncio.TimeoutError, websockets.exceptions.ConnectionClosed):
            return None, None

    async def register(self, account: str, password: str):
        await self.send(MsgId.C2S_Register, account=account, password=password)
        msg_id, msg = await self.recv()
        assert msg_id == MsgId.S2C_RegisterResult, f"Expected RegisterResult, got {msg_id}"
        return msg

    async def login(self, account: str, password: str):
        await self.send(MsgId.C2S_Login, account=account, password=password)
        msg_id, msg = await self.recv()
        assert msg_id == MsgId.S2C_LoginResult, f"Expected LoginResult, got {msg_id}"
        if msg.code == 0:
            self.uid = msg.uid
        return msg

    async def ping(self):
        ts = int(time.time() * 1000)
        await self.send(MsgId.C2S_Ping, timestamp=ts)
        msg_id, msg = await self.recv()
        assert msg_id == MsgId.S2C_Pong, f"Expected Pong, got {msg_id}"
        return msg


# â”€â”€ æ—¥å¿— â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
def log(tag: str, msg: str):
    ts = time.strftime("%H:%M:%S")
    print(f"  [{ts}] {tag}: {msg}")


def section(title: str):
    print(f"\n{'='*60}")
    print(f"  {title}")
    print(f"{'='*60}")


def ok(msg: str):
    print(f"  âœ… {msg}")


def fail(msg: str):
    print(f"  âŒ {msg}")


# â”€â”€ æµ‹è¯•ç”¨ä¾‹ â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€

async def test_register(uri: str):
    """æµ‹è¯•æ³¨å†Œæµç¨‹"""
    section("TEST: Register")
    c = GameClient(uri, "Register")
    try:
        await c.connect()

        # 1) æ³¨å†Œæ–°è´¦å·
        account = f"testuser_{int(time.time())}"
        result = await c.register(account, "pass123")
        assert result.code == 0, f"Register failed with code {result.code}"
        assert result.uid > 0, f"Invalid uid: {result.uid}"
        ok(f"æ³¨å†ŒæˆåŠŸ: account={account}, uid={result.uid}")

    except Exception as e:
        fail(f"æ³¨å†Œæµ‹è¯•å¤±è´¥: {e}")
        traceback.print_exc()
    finally:
        await c.close()


async def test_register_duplicate(uri: str):
    """æµ‹è¯•é‡å¤æ³¨å†Œ"""
    section("TEST: Register Duplicate")
    account = f"dup_user_{int(time.time())}"

    c1 = GameClient(uri, "Reg1")
    try:
        await c1.connect()
        result = await c1.register(account, "pass123")
        assert result.code == 0
        ok(f"é¦–æ¬¡æ³¨å†ŒæˆåŠŸ: uid={result.uid}")
    finally:
        await c1.close()

    # æ³¨å†ŒæˆåŠŸå gate ä¼šæ–­å¼€è¿æ¥(è®¤è¯å¤±è´¥åœºæ™¯), é‡æ–°è¿æ¥å†è¯•
    await asyncio.sleep(0.3)

    c2 = GameClient(uri, "Reg2")
    try:
        await c2.connect()
        result = await c2.register(account, "pass456")
        assert result.code == 3, f"Expected code=3 (duplicate), got {result.code}"
        ok("é‡å¤æ³¨å†Œæ­£ç¡®è¿”å› code=3")
    except Exception as e:
        fail(f"é‡å¤æ³¨å†Œæµ‹è¯•å¤±è´¥: {e}")
        traceback.print_exc()
    finally:
        await c2.close()


async def test_login(uri: str):
    """æµ‹è¯•ç™»å½•æµç¨‹"""
    section("TEST: Login")
    account = f"login_user_{int(time.time())}"

    # å…ˆæ³¨å†Œ
    c1 = GameClient(uri, "RegFirst")
    try:
        await c1.connect()
        reg = await c1.register(account, "mypass")
        assert reg.code == 0
        uid = reg.uid
        ok(f"æ³¨å†Œå®Œæˆ: uid={uid}")
    finally:
        await c1.close()

    await asyncio.sleep(0.3)

    # ç™»å½•
    c2 = GameClient(uri, "Login")
    try:
        await c2.connect()
        result = await c2.login(account, "mypass")
        assert result.code == 0, f"Login failed: code={result.code}"
        assert result.uid == uid, f"UID mismatch: expected {uid}, got {result.uid}"
        ok(f"ç™»å½•æˆåŠŸ: uid={result.uid}")
    except Exception as e:
        fail(f"ç™»å½•æµ‹è¯•å¤±è´¥: {e}")
        traceback.print_exc()
    finally:
        await c2.close()


async def test_login_wrong_password(uri: str):
    """æµ‹è¯•é”™è¯¯å¯†ç """
    section("TEST: Login Wrong Password")
    account = f"wrong_pw_{int(time.time())}"

    c1 = GameClient(uri, "RegWP")
    try:
        await c1.connect()
        reg = await c1.register(account, "correct")
        assert reg.code == 0
    finally:
        await c1.close()

    await asyncio.sleep(0.3)

    c2 = GameClient(uri, "WrongPW")
    try:
        await c2.connect()
        result = await c2.login(account, "wrong_password")
        assert result.code == 2, f"Expected code=2 (wrong password), got {result.code}"
        ok("é”™è¯¯å¯†ç æ­£ç¡®è¿”å› code=2")
    except Exception as e:
        fail(f"é”™è¯¯å¯†ç æµ‹è¯•å¤±è´¥: {e}")
        traceback.print_exc()
    finally:
        await c2.close()


async def test_login_nonexistent(uri: str):
    """æµ‹è¯•ä¸å­˜åœ¨çš„è´¦å·"""
    section("TEST: Login Nonexistent Account")
    c = GameClient(uri, "NoAcct")
    try:
        await c.connect()
        result = await c.login(f"no_such_account_{time.time()}", "whatever")
        assert result.code == 1, f"Expected code=1 (not found), got {result.code}"
        ok("ä¸å­˜åœ¨è´¦å·æ­£ç¡®è¿”å› code=1")
    except Exception as e:
        fail(f"ä¸å­˜åœ¨è´¦å·æµ‹è¯•å¤±è´¥: {e}")
        traceback.print_exc()
    finally:
        await c.close()


async def test_heartbeat(uri: str):
    """æµ‹è¯•å¿ƒè·³"""
    section("TEST: Heartbeat (Ping/Pong)")
    account = f"hb_user_{int(time.time())}"

    # æ³¨å†Œ+ç™»å½•
    c = GameClient(uri, "HB")
    try:
        await c.connect()
        reg = await c.register(account, "pass")
        assert reg.code == 0
    finally:
        await c.close()

    await asyncio.sleep(0.3)

    c = GameClient(uri, "HB")
    try:
        await c.connect()
        result = await c.login(account, "pass")
        assert result.code == 0

        # å‘é€3æ¬¡å¿ƒè·³
        for i in range(3):
            pong = await c.ping()
            assert pong.timestamp > 0
            ok(f"Ping/Pong #{i+1}: timestamp={pong.timestamp}")
            await asyncio.sleep(0.2)

    except Exception as e:
        fail(f"å¿ƒè·³æµ‹è¯•å¤±è´¥: {e}")
        traceback.print_exc()
    finally:
        await c.close()


async def test_heartbeat_before_login(uri: str):
    """æµ‹è¯•æœªç™»å½•æ—¶å‘å¿ƒè·³(åº”è¢«å¿½ç•¥ï¼Œä¸å´©æºƒ)"""
    section("TEST: Heartbeat Before Login")
    c = GameClient(uri, "HBNoAuth")
    try:
        await c.connect()

        # æœªè®¤è¯å°±å‘ ping â€” gate åº”è¯¥ä¸è½¬å‘(uid=nil), æ— å›å¤
        ts = int(time.time() * 1000)
        await c.send(MsgId.C2S_Ping, timestamp=ts)

        msg_id, msg = await c.recv_optional(timeout=2.0)
        if msg_id is None:
            ok("æœªè®¤è¯çš„å¿ƒè·³è¢«å¿½ç•¥(æ— å“åº”)ï¼Œç¬¦åˆé¢„æœŸ")
        else:
            # æŸäº›å®ç°å¯èƒ½ä»ç„¶å›å¤ pong, ä¸ç®—é”™è¯¯
            ok(f"æœªè®¤è¯å¿ƒè·³æ”¶åˆ°å“åº”: {MSGID_NAME.get(msg_id, msg_id)} (å¯æ¥å—)")

    except Exception as e:
        fail(f"æœªç™»å½•å¿ƒè·³æµ‹è¯•å¤±è´¥: {e}")
        traceback.print_exc()
    finally:
        await c.close()


async def test_reconnect(uri: str):
    """æµ‹è¯•é¡¶å·: åŒä¸€è´¦å·ä»å¦ä¸€ä¸ªè¿æ¥ç™»å½•, æ—§è¿æ¥åº”è¢«è¸¢"""
    section("TEST: Reconnect / Kick Duplicate Login")
    account = f"recon_user_{int(time.time())}"

    # æ³¨å†Œ
    c0 = GameClient(uri, "RegRecon")
    try:
        await c0.connect()
        reg = await c0.register(account, "pass")
        assert reg.code == 0
        uid = reg.uid
    finally:
        await c0.close()

    await asyncio.sleep(0.3)

    # ç¬¬ä¸€æ¬¡ç™»å½•
    c1 = GameClient(uri, "Old")
    try:
        await c1.connect()
        r1 = await c1.login(account, "pass")
        assert r1.code == 0
        ok(f"ç¬¬ä¸€æ¬¡ç™»å½•æˆåŠŸ: uid={r1.uid}")

        # ç¬¬äºŒæ¬¡ç™»å½•(é¡¶å·)
        c2 = GameClient(uri, "New")
        try:
            await c2.connect()
            r2 = await c2.login(account, "pass")
            assert r2.code == 0
            ok(f"ç¬¬äºŒæ¬¡ç™»å½•æˆåŠŸ: uid={r2.uid}")

            # æ—§è¿æ¥åº”æ”¶åˆ° Kick æˆ–è¢«æ–­å¼€
            await asyncio.sleep(0.5)
            kick_id, kick_msg = await c1.recv_optional(timeout=2.0)
            if kick_id == MsgId.S2C_Kick:
                ok(f"æ—§è¿æ¥æ”¶åˆ° Kick: reason={kick_msg.reason}")
            elif kick_id is None:
                ok("æ—§è¿æ¥å·²è¢«æ–­å¼€(è¿æ¥å…³é—­)")
            else:
                log("Old", f"æ”¶åˆ°æ„å¤–æ¶ˆæ¯: {MSGID_NAME.get(kick_id, kick_id)}")

            # æ–°è¿æ¥åº”æ­£å¸¸å·¥ä½œ
            pong = await c2.ping()
            ok(f"æ–°è¿æ¥å¿ƒè·³æ­£å¸¸: timestamp={pong.timestamp}")

        finally:
            await c2.close()

    except websockets.exceptions.ConnectionClosed:
        ok("æ—§è¿æ¥è¢«æœåŠ¡ç«¯å…³é—­(ç¬¦åˆé¢„æœŸ)")
    except Exception as e:
        fail(f"é¡¶å·æµ‹è¯•å¤±è´¥: {e}")
        traceback.print_exc()
    finally:
        try:
            await c1.close()
        except Exception:
            pass


async def test_stress(uri: str, count: int = 20):
    """å¹¶å‘å¤šè¿æ¥å‹åŠ›æµ‹è¯•"""
    section(f"TEST: Stress ({count} concurrent connections)")

    results = {"ok": 0, "fail": 0}

    async def single_client(index: int):
        account = f"stress_{int(time.time())}_{index}"
        c = GameClient(uri, f"S{index:03d}")
        try:
            await c.connect()
            reg = await c.register(account, "p")
            if reg.code != 0:
                results["fail"] += 1
                return
            await c.close()

            await asyncio.sleep(0.1)

            await c.connect()
            login = await c.login(account, "p")
            if login.code != 0:
                results["fail"] += 1
                return

            pong = await c.ping()
            assert pong.timestamp > 0

            results["ok"] += 1
        except Exception as e:
            log(f"S{index:03d}", f"error: {e}")
            results["fail"] += 1
        finally:
            try:
                await c.close()
            except Exception:
                pass

    # åˆ†æ‰¹æ‰§è¡Œ, æ¯æ‰¹ 5 ä¸ª
    batch_size = 5
    for start in range(0, count, batch_size):
        batch = [single_client(i) for i in range(start, min(start + batch_size, count))]
        await asyncio.gather(*batch)
        await asyncio.sleep(0.2)

    ok(f"å‹åŠ›æµ‹è¯•å®Œæˆ: æˆåŠŸ={results['ok']}, å¤±è´¥={results['fail']}")
    if results["fail"] > 0:
        fail(f"{results['fail']} ä¸ªè¿æ¥å¤±è´¥")


async def test_join_room(uri: str):
    """æµ‹è¯•åŠ å…¥æˆ¿é—´(è·¨æœç©æ³•)"""
    section("TEST: Join Room")
    account = f"room_user_{int(time.time())}"

    # æ³¨å†Œ
    c0 = GameClient(uri, "RegRoom")
    try:
        await c0.connect()
        reg = await c0.register(account, "pass")
        assert reg.code == 0
    finally:
        await c0.close()

    await asyncio.sleep(0.3)

    c = GameClient(uri, "Room")
    try:
        await c.connect()
        result = await c.login(account, "pass")
        assert result.code == 0
        ok(f"ç™»å½•æˆåŠŸ: uid={result.uid}")

        # åŠ å…¥æˆ¿é—´
        await c.send(MsgId.C2S_JoinRoom, roomId="test_room_001")
        msg_id, msg = await c.recv(timeout=3.0)

        if msg_id == MsgId.S2C_JoinResult:
            if msg.code == 0:
                ok(f"åŠ å…¥æˆ¿é—´æˆåŠŸ: roomId={msg.roomId}")
            else:
                log("Room", f"åŠ å…¥æˆ¿é—´è¿”å› code={msg.code}")
        else:
            log("Room", f"æ”¶åˆ°æ„å¤–å“åº”: {MSGID_NAME.get(msg_id, msg_id)}")

    except asyncio.TimeoutError:
        log("Room", "ç­‰å¾… JoinResult è¶…æ—¶(å¯èƒ½æ¨¡å—æœªæ³¨å†Œè·¯ç”±)")
    except Exception as e:
        fail(f"æˆ¿é—´æµ‹è¯•å¤±è´¥: {e}")
        traceback.print_exc()
    finally:
        await c.close()


# â”€â”€ æµ‹è¯•ç¼–æ’ â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€

TEST_REGISTRY = {
    "register":     [test_register, test_register_duplicate],
    "login":        [test_login, test_login_wrong_password, test_login_nonexistent],
    "heartbeat":    [test_heartbeat, test_heartbeat_before_login],
    "reconnect":    [test_reconnect],
    "room":         [test_join_room],
    "stress":       [test_stress],
}


async def run_tests(uri: str, test_name: str):
    print(f"\nğŸ® Skynet Game Server Test â€” target: {uri}")
    print(f"   test suite: {test_name}")

    if test_name == "all":
        for group_name, tests in TEST_REGISTRY.items():
            for t in tests:
                await t(uri)
                await asyncio.sleep(0.3)
    elif test_name in TEST_REGISTRY:
        for t in TEST_REGISTRY[test_name]:
            await t(uri)
            await asyncio.sleep(0.3)
    else:
        print(f"âŒ Unknown test: '{test_name}'")
        print(f"   Available: {', '.join(TEST_REGISTRY.keys())}, all")
        return

    section("DONE")
    print("  All tests completed.\n")


def main():
    parser = argparse.ArgumentParser(description="Skynet Game Server Test Client")
    parser.add_argument("--host", default="127.0.0.1", help="Server host (default: 127.0.0.1)")
    parser.add_argument("--port", type=int, default=9948, help="WebSocket port (default: 9948)")
    parser.add_argument("--test", default="all",
                        help="Test to run: register, login, heartbeat, reconnect, room, stress, all")
    args = parser.parse_args()

    uri = f"ws://{args.host}:{args.port}"
    asyncio.run(run_tests(uri, args.test))


if __name__ == "__main__":
    main()
